Add standard RK4 timestepper

geoffrey4444 opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

Note: no support for dense output, not stability preserving.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,187 @@
+      // On entry V = v^(3), U0 = u^n, rhs0 = RHS(v^(3), t^n + dt),
+      // time = t^n + (1/2)*dt
+      *u = U0 + time_step.value() * dt_vars;
+      // On exit v = v^(4), time = t^n + dt
```
Can update `*u` directly here to get the final step by moving the combining code below into the `case 3` section.

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,187 @@
+}
+
+template <typename Vars, typename DerivVars>
+void RungeKutta4::dense_update_u(gsl::not_null<Vars*> /* u */,
```
`const not_null`

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.cpp:
```diff
@@ -0,0 +1,66 @@
+size_t RungeKutta4::number_of_past_steps() const noexcept { return 0; }
+
+double RungeKutta4::stable_step() const noexcept {
+  ERROR("Unimplemented");
```
this should just return `1.0` since that's the stability radius of the stepper.

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.cpp:
```diff
@@ -0,0 +1,66 @@
+             "Wrong substep time");
+      return {current_id.time_runs_forward(), current_id.slab_number(),
+              current_id.step_time(), 4, current_id.step_time() + time_step};
+    case 4:
```
Don't need `case 4` anymore

geoffrey4444 replied:
Except my workaround for dense output does need that case...unless there's a better way...

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.cpp:
```diff
@@ -0,0 +1,66 @@
+
+namespace TimeSteppers {
+
+uint64_t RungeKutta4::number_of_substeps() const noexcept { return 5; }
```
change to `4`

geoffrey4444 replied:
My current workaround actually made me change this back to 5, since I added an extra substep to get dense output working.

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,187 @@
+      // v^(1) = u^n + dt*RHS(u^n,t^n)/2
+      // On entry V = u^n, U0 = u^n, rhs0 = RHS(u^n, t^n),
+      // time = t^n
+      *u += (time_step.value() * dt_vars) / 2;
```
better to multiply by `0.5` and everywhere

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,187 @@
+      *u = (2 * (history->begin() + 1).value() +
+            4 * (history->begin() + 2).value() +
+            2 * (history->begin() + 3).value() +
+            (history->begin() + 4).value() - 3 * U0) /
```
You can do `* (1.0 / 6.0)` and then the compiler will generally do the compilation for you. If you want to be extra sure you can do:
```cpp
constexpr double one_sixth = 1.0 / 6.0;
*u = (...) * one_sixth;
```

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,187 @@
+      // u^(n+1) = (2v^(1) + 4*v^(2) + 2*v^(3) + v^(4) - 3*U0)/6
+      // On entry V = v^(4), U0 = u^n, rhs0 = RHS(v^(4), t^n + dt),
+      // time = t^n + dt
+      *u = (2 * (history->begin() + 1).value() +
```
style wise `2` -> `2.0`, etc.

nilsdeppe commented on tests/Unit/Time/TimeSteppers/Test_RungeKutta4.cpp:
```diff
@@ -0,0 +1,26 @@
+SPECTRE_TEST_CASE("Unit.Time.TimeSteppers.RungeKutta4", "[Unit][Time]") {
+  const TimeSteppers::RungeKutta4 stepper{};
+  TimeStepperTestUtils::check_substep_properties(stepper);
+  TimeStepperTestUtils::integrate_test(stepper, 0, 1., 1e-9);
```
`1.e-9`, etc.

nilsdeppe commented on tests/Unit/Time/TimeSteppers/Test_RungeKutta4.cpp:
```diff
@@ -0,0 +1,26 @@
+  TimeStepperTestUtils::integrate_test(stepper, 0, 1., 1e-9);
+  TimeStepperTestUtils::integrate_test(stepper, 0, -1., 1e-9);
+  TimeStepperTestUtils::integrate_variable_test(stepper, 0, 1e-9);
+  TimeStepperTestUtils::check_convergence_order(stepper, 4);
```
Please add a line testing stable step as well.

nilsdeppe reviewed (changes requested):
Review with @wthrowe (primary) @markscheel @fmahebert @moxcodes @geoffrey4444

geoffrey4444 committed:
Add standard RK4 timestepper

geoffrey4444 commented:
@wthrowe this is ready for you to take a look. My apologies for not doing a fixup commit! I meant to but accidentally messed up git :(, so that I overwrote the old version on GitHub but only had the version on my computer with all the fixups already made, so there was no easy way to split the changes back out into a fixup commit. Hopefully because this PR is short it's not too big a hassle. I'll be more careful next time!

@wthrowe @nilsdeppe please especially take a look at stable_step()? I think the discussion yesterday was wrong, saying it should return 1.0. The stability region of RK4 is not a disk, like it is for Euler, and in fact when I tried it, the stability test showed dy/dt = -2 y was stable for dt=1.1. Trial and error suggested it dt=1.4 roughly is the biggest stable step for dy/dt=-2y. I worked out the correct stable step is actually about dt=1.39265, as discussed in the comments for stable_step(), so I now have stable_step return 1.39265 instead of 1.0.

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,184 @@
+ * v^{(2)} & = & u^n + dt\cdot RHS(t^n + dt/2, v^{(1)})/2,\\
+ * v^{(3)} & = & u^n + dt\cdot RHS(t^n + dt/2, v^{(2)}),\\
+ * v^{(4)} & = & u^n + dt\cdot RHS(t^n + dt, v^{(3)}),\\
+ * v^{(5)} = u^{n+1} & = & (2v^{(1)} + 4v^{(2)} + 2v^{(3)} + v^{(4)} - 3 u^n)/6.
```
Should we remove `v^{(5)}`?

wthrowe commented:
> I think the discussion yesterday was wrong, saying it should return 1.0. The stability region of RK4 is not a disk, like it is for Euler, and in fact when I tried it, the stability test showed dy/dt = -2 y was stable for dt=1.1. Trial and error suggested it dt=1.4 roughly is the biggest stable step for dy/dt=-2y. I worked out the correct stable step is actually about dt=1.39265, as discussed in the comments for stable_step(), so I now have stable_step return 1.39265 instead of 1.0.

There are a number of different maximum "stable steps," depending on what you are integrating.  With AB and RK3 (and I'm guessing RK4 as well, though I haven't checked) the ones we care about happen to have the same values.  The stability reporting code may have to get more sophisticated in the future.  What you have done is what I want for the time being.

wthrowe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,184 @@
+ * v^{(5)} = u^{n+1} & = & (2v^{(1)} + 4v^{(2)} + 2v^{(3)} + v^{(4)} - 3 u^n)/6.
+ * \f}
+ *
+ * Note that in the implementation, the expressio nfor \f$v^{(4)}\f$ is
```
Just noticed, there's a character transposition here (` n` -> `n `).  Very minor, so I'm fine with ignoring it.

nilsdeppe commented:
Since @wthrowe and I are both okay deferring the minor typos and this still needs dense output let's merge. 

@geoffrey4444 could you minor changes to the docs in the dense output PR? Thanks!

nilsdeppe commented:
Oops, can't merge, this fails to build. @geoffrey4444 could you fix that please and also the two minor comments?

geoffrey4444 commented:
Yep! I will fix...had a long wait for Travis...will fix docs and Travis ASAP...thanks!

geoffrey4444 commented:
Pushed, squashed since changes were so small (the 2 doc changes + an unused variable warning that only came up in gcc, not clang, so I missed it). I'll fix any other travis errors once it runs again, then let you know when it's ready to look at again

geoffrey4444 commented:
Travis failed again, this time gcc 6 and 7 (but not clang) wanted my unimplemented function marked [[noreturn]]. Just pushed and squashed this fix...we'll see if travis passes this time

geoffrey4444 commented:
gcc release builds still giving no return warnings...maybe I should just implement dense output, because it would be easier than fixing these warnings

nilsdeppe commented:
I agree, especially given that RK4 dense output is relatively straightforward :)

geoffrey4444 commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,199 @@
+                                 const History<Vars, DerivVars>& /* history */,
+                                 const double /* time */) const noexcept {
+  ERROR("Unimplemented");
+}
```
Why not just remove this function for now so we can merge, then implement dense output?

nilsdeppe replied:
I'm confused... Are you asking yourself this question or is github mixing usernames? I think implementing it was first mentioned by @geoffrey4444 in this comment: https://github.com/sxs-collaboration/spectre/pull/1790#issuecomment-555536117

geoffrey4444 replied:
Yes, that was a note to myself...irrelevant now that I implemented dense output

wthrowe commented:
I'm making a change to the TimeStepper interface in #1828.  Whichever PR is merged later will have to add an implementation of `can_change_step_size` for `RungeKutta4`.  The implementation for RK4 is the same as for RK3, i.e.,
```c++
  template <typename Vars, typename DerivVars>
  bool can_change_step_size(
      const TimeStepId& time_id,
      const TimeSteppers::History<Vars, DerivVars>& /*history*/) const
      noexcept {
    return time_id.substep() == 0;
  }
```
For consistency, it should appear immediately after the declaration of `next_time_id`.  Testing for this method is being added to `TimeStepperTestUtils::check_substep_properties`, so no change to the test is required.

geoffrey4444 commented:
This is ready for another look @wthrowe and @nilsdeppe. I added dense output as a fixup commit. One issue I ran into is that in the dense output test, dense output only is passed history containing all substeps except for the final one, while RK4 dense output needs the results of the final substep. I worked around this by adding a do-nothing extra substep, but if there's a better workaround, I'd love to know a better way to do this!

wthrowe commented:
Approximating the derivative at the endpoint by the derivative at the last substep should be sufficiently accurate.  (I believe they differ by O(dt^3).)

geoffrey4444 commented:
@wthrowe thanks for the quick comment! I’m surprised if you’re right that that’s good enough, because I started out implementing dense output as you suggest, using the last substep instead of the final result, but the dense output test failed. This is the only way I could find to make that test pass.

wthrowe commented:
I did a hacky test (without actually removing the last substep), and it seems to work.  I changed the 4 in `(history.begin() + 4).derivative()` to a 3, and everything still passes.  The "4.value()" expressions still need to be eliminated, but they are algebraically related to the quantities from the earlier substeps so that shouldn't be an issue.

As a separate note: #1828 has now been merged, so you will need to add the code I gave above.

geoffrey4444 commented:
@wthrowe I just pushed a mixup that implements your idea for dense output of approximating the derivative and recomputing algebraically the value at the final step. I also cleaned up the docs and added the function from #1828. I think it's ready for another look when you get a chance!

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.cpp:
```diff
@@ -0,0 +1,68 @@
+  switch (current_id.substep()) {
+    case 0:
+      ASSERT(current_id.substep_time() == current_id.step_time(),
+             "Wrong substep time");
```
Could you print the given time and the expected time? I've personally found errors that don't provide that sort of info rather frustrating :)

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,218 @@
+ * \f$u(t^{n+1})=u^{n+1}\f$ using the following equations:
+ *
+ * \f{eqnarray}{
+ * v^{(1)} & = & u^n + dt\cdot RHS(t^n, u^n)/2,\\
```
[optional] In a decent amount of literature (the one pertaining to timestepping in DG/FV/FD specifically) I've seen `\mathcal{L}` used instead of `RHS`

wthrowe replied:
The LTS literature doesn't seem to have a consistent notation for this quantity, so `\mathcal{L}` seems as good as anything from my side.

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,218 @@
+  constexpr double one_sixth = 1.0 / 6.0;
+  const auto& U0 = history.begin().value();
+  const auto& dt_vars = (history.end() - 1).derivative();
+  const auto& U1 =
```
Why is this a reference? I'm also worried that this will create an expression template when using `auto` and so you are recomputing this expression twice below (if I counted correctly)

geoffrey4444 replied:
I can't think of a reason not to just change to const auto u1

nilsdeppe replied:
I think `const auto u1` will mean the type is an expression and so would be re-evaluated twice below rather than computed once and stored. Does that make sense?

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,218 @@
+                                 const History<Vars, DerivVars>& history,
+                                 const double time) const noexcept {
+  ASSERT(history.size() == number_of_substeps(),
+         "Can only dense output on last substep");
```
Can you give the current substep in the error message and the total number of substeps?

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,218 @@
+  const double step_end = history[history.size() - 1].value();
+  const double time_step = step_end - step_start;
+  const double output_fraction = (time - step_start) / time_step;
+  ASSERT(output_fraction >= 0, "Attempting dense output at time "
```
`0` -> `0.0`

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,218 @@
+  ASSERT(output_fraction >= 0, "Attempting dense output at time "
+                                   << time << ", but already progressed past "
+                                   << step_start);
+  ASSERT(output_fraction <= 1, "Requested time ("
```
`1` -> `1.0`

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,218 @@
+  // from previous substeps, and ii) the derivative at the final step is
+  // approximated as the derivative at the penultimate substep.
+  constexpr double one_sixth = 1.0 / 6.0;
+  const auto& U0 = history.begin().value();
```
`u0` to be consistent with lower case `u` (same with `u1`)

nilsdeppe reviewed (changes requested):
I'm happy for you to squash once @wthrowe has had a chance to look at the commits.

wthrowe commented on src/Time/TimeSteppers/RungeKutta4.hpp:
```diff
@@ -0,0 +1,218 @@
+      // v^(1) = u^n + dt*RHS(u^n,t^n)/2
+      // On entry V = u^n, U0 = u^n, rhs0 = RHS(u^n, t^n),
+      // time = t^n
+      *u += 0.5 * (time_step.value() * dt_vars);
```
Remove the extra parens here (and in other similar expressions).  `double * double * DataVector` (equivalently, `(double * double) * DataVector`) is faster than `double * (double * DataVector)`.

wthrowe reviewed (changes requested):
I only have one thing to add to the above.

wthrowe reviewed (approved):
Travis failed in some very strange way.  I'm pretty sure it's not related to the changes here, but someone who understands that setup better should take a look.

nilsdeppe commented:
Travis has started doing something weird with their hardware again, I think. About a week ago these errors started cropping up... I'm reasonably confident in GH actions at this point so I'll merge this.
